### PR TITLE
MAT-20 topup

### PR DESCRIPTION
**Pull request by jdkanani**

### CODE REVIEW

File: auth/ante.go [hunk 1]
```diff
@@ -26,7 +26,8 @@ var (
 	gasWantedPerCheckpoinTx sdk.Gas = 10000000
 	gasUsedPerCheckpointTx  sdk.Gas = gasWantedPerCheckpoinTx - 1000000
 
-	feeWantedPerTx = types.Coins{types.Coin{Denom: "vetic", Amount: types.NewInt(1)}}
+	// FeeWantedPerTx fee wanted per tx
+	FeeWantedPerTx = types.Coins{types.Coin{Denom: "vetic", Amount: types.NewInt(1)}}
```
Comment (vaibhavchellani) [inline auth/ante.go]: rename to `gas` to remove confusion

File: auth/ante.go [hunk 2]
```diff
@@ -100,7 +83,7 @@ func NewAnteHandler(
 
 		// gas for tx
 		gasForTx := gasWantedPerTx // stdTx.Fee.Gas
-		feeForTx := feeWantedPerTx // stdTx.Fee.Amount
+		feeForTx := FeeWantedPerTx // stdTx.Fee.Amount
```
Comment (vaibhavchellani) [inline auth/ante.go]: Why create this variable at all?

File: bank/handler.go [hunk 3]
```diff
@@ -51,3 +59,96 @@ func handleMsgMultiSend(ctx sdk.Context, k Keeper, msg types.MsgMultiSend) sdk.R
 		Tags: tags,
 	}
 }
+
+// Handle MsgMintFeeToken
+func handleMsgTopup(ctx sdk.Context, k Keeper, msg types.MsgTopup, contractCaller helper.IContractCaller) sdk.Result {
+	if !k.GetSendEnabled(ctx) {
```
Comment (vaibhavchellani) [inline bank/handler.go]: Do we need `sendEnabled` or `sendDisabled` type functionality in `heimdall` ?

File: bank/keeper.go [hunk 4]
```diff
@@ -105,7 +119,48 @@ func (keeper BaseKeeper) InputOutputCoins(
 	return inputOutputCoins(ctx, keeper.ak, inputs, outputs)
 }
 
-// between accounts without the possibility of creating coins.
+// GetTopup returns validator toptup
```
Comment (vaibhavchellani) [inline bank/keeper.go]:
```suggestion
// GetTopup returns validator toptup information
```

File: bank/keeper.go [hunk 5]
```diff
@@ -105,7 +119,48 @@ func (keeper BaseKeeper) InputOutputCoins(
...
+
+	// validator topup
+	bz, err := bankTypes.MarshallValidatorTopup(keeper.cdc, validatorTopup)
+	if err != nil {
+		return err
+	}
+
+	// store validator with address prefixed with validator key as index
+	store.Set(GetTopupKey(addr.Bytes()), bz)
+	keeper.Logger(ctx).Debug("Validator topup stored", "key", hex.EncodeToString(GetTopupKey(addr.Bytes())))
```
Comment (vaibhavchellani) [inline bank/keeper.go]: Also log the amount here

File: bank/types/codec.go [hunk 6]
```diff
@@ -10,12 +10,14 @@ import (
 func RegisterCodec(cdc *codec.Codec) {
 	cdc.RegisterConcrete(MsgSend{}, "cosmos-sdk/MsgSend", nil)
 	cdc.RegisterConcrete(MsgMultiSend{}, "cosmos-sdk/MsgMultiSend", nil)
+	cdc.RegisterConcrete(MsgTopup{}, "cosmos-sdk/MsgTopup", nil)
```
Comment (vaibhavchellani) [inline bank/types/codec.go]:
```suggestion
        cdc.RegisterConcrete(MsgSend{}, "bank/MsgSend", nil)
	cdc.RegisterConcrete(MsgMultiSend{}, "bank/MsgMultiSend", nil)
	cdc.RegisterConcrete(MsgTopup{}, "bank/MsgTopup", nil)
```

File: auth/ante.go (hunk 2 above)
Comment (jdkanani) [reply auth/ante.go]: Just to create logical block, nothing else.

File: auth/ante.go (hunk 1 above)
Comment (jdkanani) [reply auth/ante.go]: That's not gas. That's fee per tx only. Gas needs price to get actual fee per tx.

**Comment by vaibhavchellani** [comment]
@jdkanani please checkout comments, post resolution, we can merge this!

File: bank/handler.go (hunk 3 above)
Comment (jdkanani) [reply bank/handler.go]: Yes, we have. Governance module handles that for future use.

File: auth/ante.go (hunk 2 above)
Comment (vaibhavchellani) [reply auth/ante.go]: Cool

**Pull request merged**